### ISSUE TITLE
Support multiple participants

### ISSUE THREAD

**Issue by jonnor** [msgflo/msgflo-arduino#2]
A participant should be able to model a particular "service", as as group of functionally-related ports. Independent pieces should be independent participants. A device should be able to "host" many of these.

Some preparation has been done in the code for this, though the `Participant` class - but Engine implementation needs to support more than one.

**Comment by jonnor** [comment]
An API sketch here, also simplifying Engine setup and registering the particpant/ports.

```

#include <string>
#include <vector>
#include <functional>

struct InPort {
    std::string id;
    std::string queue;
    std::string type = "any"; //default
    std::function< void (const uint8_t*, size_t) > callback;
};

struct Publisher {
    virtual bool send(const uint8_t *data, size_t len) = 0;
};

struct OutPort {
    std::string id;
    std::string queue;
    std::string type = "any"; //default

    Publisher *publisher;

    bool send(const uint8_t *data, size_t len) {
        if (publisher) {
            return publisher->send(data, len);
        } else {
            // fail, we're probably not initialized yet
            return false;
        }
    };
};

struct Participant {
    std::string component;
    std::vector<InPort *> inports;
    std::vector<OutPort *> outports;
};

struct PrintPublisher : public Publisher {
    bool send(const uint8_t *data, size_t len) {
        fprintf(stderr, "%.*s\n", (int)len, data);
        return true;
    }
};

struct Engine {
    PrintPublisher print;

    bool add(const Participant &pa) {
        for (auto &p : pa.outports) {
            p->publisher = &print;
        }
        for (auto &p : pa.inports) {
            // test roundtrip
            const uint8_t data[] = "abc";
            p->callback(data, sizeof(data));
        }
        return true;
    }
};

// App code

struct Config {
    const std::string id = "iddd";
} config;

struct MyThing {
    InPort in {
        id: "in",
        queue: "foo/" + config.id + "/bar",
    };
    OutPort out {
        id: "out",
        queue: "foo/" + config.id + "/barout",
    };
    const Participant part {
        component: "my/Foo",
        inports: { &in },
        outports: { &out },
    };
};


MyThing thing;
Engine engine;

int main() {
    thing.in.callback = [](const uint8_t *data, size_t len) {
        thing.out.send(data, len);
    };
    engine.add(thing.part);

    thing.out.send((uint8_t*)"12", 2);
    return 0;
}
```

**Comment by jonnor** [comment]
Fixed in #14
